client/src/message-handler.js

```diff
@@ -2,280 +2,523 @@ var SyncMessage = require('../../lib/syncmessage');
-    // with their checksums to version them
-    async.eachSeries(syncedPaths, stampChecksum, function(err) {
+    fs.lstat(syncedPath, function(err, stats) {
       if(err) {
-        return onError(syncManager, err);
+        if(err.code !== 'ENOENT') {
+          log.error('Failed to access ' + syncedPath + ' in handlePatchAckResponse');
+          return fs.delaySync(function(delayErr, delayedPath) {
+            if(delayErr) {
+              log.error('Failed to delay upstream sync for ' + delayedPath + ' in handlePatchAckResponse');
+            }
+            onError(syncManager, err);
+          });
+        }
+
+        // Non-existent paths (usually due to renames or
+        // deletes cannot be stamped with a checksum
+        return complete();
       }
 
-      sync.onCompleted(data.content.syncedPaths);
+      if(!stats.isFile()) {
+        return complete();
+      }
+
+      rsyncUtils.getChecksum(rawFs, syncedPath, function(err, checksum) {
+        if(err) {
+          log.error('Failed to get the checksum for ' + syncedPath + ' in handlePatchAckResponse');
+          return fs.delaySync(function(delayErr, delayedPath) {
+            if(delayErr) {
+              log.error('Failed to delay upstream sync for ' + delayedPath + ' in handlePatchAckResponse while getting checksum');
+            }
+            onError(syncManager, err);
+          });
+        }
+
+        fsUtils.setChecksum(rawFs, syncedPath, checksum, function(err) {
+          if(err) {
+            log.error('Failed to stamp the checksum for ' + syncedPath + ' in handlePatchAckResponse');
+            return fs.delaySync(function(delayErr, delayedPath) {
+              if(delayErr) {
+                log.error('Failed to delay upstream sync for ' + delayedPath + ' in handlePatchAckResponse while setting checksum');
+              }
+              onError(syncManager, err);
+            });
+          }
+
+          complete();
+        });
+      });
     });
   }
 
-  function handlePatchResponse() {
-    var modifiedPath = fs.modifiedPath;
-    fs.modifiedPath = null;
+  function handleDiffResponse() {
+    var message;
 
-    // If there was a change to the filesystem that shares a common path with
-    // the path being synced, regenerate the checksums and send them
-    // (even if it is the initial one)
-    if(modifiedPath && hasCommonPath(session.path, modifiedPath)) {
-      return resendChecksums();
+    if(data.invalidContent(['type', 'diffs'])) {
+      log.error('Path, type or diffs not sent by server in handleDiffResponse.', data);
+      return onError(syncManager, new Error('Server sent insufficient content'));
     }
 
-    var diffs = data.content.diffs;
-    diffs = deserializeDiff(diffs);
+    var path = data.content.path;
+    var type = data.content.type;
+    var diffs = deserializeDiff(data.content.diffs);
+    var changedDuringDownstream = fs.changesDuringDownstream.indexOf(path);
```

Comment:
I wouldn't bother with these two variables if you only use them here.  I'd do this:

```js
if(fs.changesDuringDownstream.indexOf(path) !== -1) {
  // Resend the checksums for that path
  return sendChecksums(syncManager, path, type, fs.stopRecording(path));
}
```